plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java

```diff
@@ -455,64 +455,70 @@ private DeltaLakeTransactionLogEntry buildRemoveEntry(ConnectorSession session,
-        Block block = blocks[0];
-        log.debug("Building add entry from %s pagePosition %d", block, pagePosition);
-        if (block.isNull(pagePosition)) {
-            return null;
-        }
-        RowType type = (RowType) parquetFields.get("add");
-        boolean deletionVectorsEnabled = isDeletionVectorEnabled(metadataEntry, protocolEntry);
-        SqlRow addEntryRow = block.getObject(pagePosition, SqlRow.class);
-        log.debug("Block %s has %s fields", block, addEntryRow.getFieldCount());
-        CheckpointFieldReader add = new CheckpointFieldReader(session, addEntryRow, type);
+        @Nullable
+        @Override
+        public DeltaLakeTransactionLogEntry getEntry(ConnectorSession session, int pagePosition, Block... blocks)
+        {
+            checkState(blocks.length == getRequiredBlocks(), "Unexpected amount of blocks: %s", blocks.length);
+            Block block = blocks[0];
+            log.debug("Building add entry from %s pagePosition %d", block, pagePosition);
+            if (block.isNull(pagePosition)) {
+                return null;
+            }
+            RowType type = (RowType) parquetFields.get("add");
+            boolean deletionVectorsEnabled = isDeletionVectorEnabled(metadataEntry, protocolEntry);
+            SqlRow addEntryRow = block.getObject(pagePosition, SqlRow.class);
+            log.debug("Block %s has %s fields", block, addEntryRow.getFieldCount());
+            CheckpointFieldReader add = new CheckpointFieldReader(session, addEntryRow, type);
+
+            Map<String, String> partitionValues = add.getMap(stringMap, "partitionValues");
+            Map<String, Optional<String>> canonicalPartitionValues = canonicalizePartitionValues(partitionValues);
+            if (!partitionConstraint.isAll() && !partitionMatchesPredicate(canonicalPartitionValues, partitionConstraint.getDomains().orElseThrow())) {
+                return null;
+            }
 
-        Map<String, String> partitionValues = add.getMap(stringMap, "partitionValues");
-        Map<String, Optional<String>> canonicalPartitionValues = canonicalizePartitionValues(partitionValues);
-        if (!partitionConstraint.isAll() && !partitionMatchesPredicate(canonicalPartitionValues, partitionConstraint.getDomains().orElseThrow())) {
-            return null;
-        }
+            String path = add.getString("path");
+            long size = add.getLong("size");
+            long modificationTime = add.getLong("modificationTime");
+            boolean dataChange = add.getBoolean("dataChange");
+
+            Optional<DeletionVectorEntry> deletionVector = Optional.empty();
+            if (deletionVectorsEnabled) {
+                deletionVector = Optional.ofNullable(add.getRow("deletionVector"))
+                        .map(row -> {
+                            RowType.Field deletionVectorField = type.getFields().stream().filter(field -> field.getName().orElseThrow().equals("deletionVector")).collect(onlyElement());
```

Comment:
Does this really need to run for every position ? Seems like type should be something we know upfront or at least per file.